sympy/matrices/expressions/matexpr.py

```diff
@@ -348,6 +349,87 @@ def canonicalize(self):
     def as_coeff_mmul(self):
         return 1, MatMul(self)
 
+    @staticmethod
+    def from_index_summation(expr, first_index):
```

Comment:
Can you add a docstring

Reply:
Why does it only need the first index?

Reply:
The last index gets detected automatically by following the summed over indices. The first and last matrices have only one summed over index, the other being either the first or last index.